src/adapter/src/catalog/apply.rs

```diff
@@ -42,6 +43,42 @@ use tracing::warn;
 
 use crate::catalog::{BuiltinTableUpdate, Catalog, CatalogState};
 
+/// Maintains the state of retractions while applying catalog state updates for a single timestamp.
+/// [`CatalogState`] maintains denormalized state for certain catalog objects. Updating an object
+/// results in applying a retraction for that object followed by applying an addition for that
+/// object. When applying those additions it can be extremely expensive to re-build that
+/// denormalized state from scratch. To avoid that issue we stash the denormalized state from
+/// retractions, so it can be used during additions.
+///
+/// Not all objects maintain denormalized state, so we only stash the retractions for the subset of
+/// objects that maintain denormalized state.
+// TODO(jkosh44) It might be simpler or more future proof to include all object types here, even if
+// the update step is a no-op for certain types.
+#[derive(Debug, Clone)]
+struct InProgressRetractions {
+    roles: BTreeMap<RoleKey, Role>,
+    databases: BTreeMap<DatabaseKey, Database>,
+    schemas: BTreeMap<SchemaKey, Schema>,
+    clusters: BTreeMap<ClusterKey, Cluster>,
+    items: BTreeMap<ItemKey, CatalogEntry>,
+    introspection_source_indexes: BTreeMap<GlobalId, CatalogEntry>,
+    system_object_mappings: BTreeMap<GlobalId, CatalogEntry>,
+}
+
+impl InProgressRetractions {
+    fn new() -> InProgressRetractions {
```

Comment:
maybe add `Default` to `derive` instead?

Reply:
Good call, done.